core/optaplanner-constraint-streams/src/test/java/org/optaplanner/constraint/streams/bavet/NodeSharingTest.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.constraint.streams.bavet;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.function.BiPredicate;
+import java.util.function.Predicate;
+
+import org.junit.jupiter.api.Test;
+import org.optaplanner.core.api.score.stream.bi.BiConstraintStream;
+import org.optaplanner.core.api.score.stream.uni.UniConstraintStream;
+import org.optaplanner.core.impl.testdata.domain.TestdataEntity;
+import org.optaplanner.core.impl.testdata.domain.TestdataSolution;
+
+class NodeSharingTest {
+
+    @Test
+    void nodeSharingForEach() {
+        BavetConstraintFactory<TestdataSolution> constraintFactory =
+                new BavetConstraintFactory<>(TestdataSolution.buildSolutionDescriptor());
+        UniConstraintStream<TestdataEntity> stream1 = constraintFactory.forEach(TestdataEntity.class);
+        UniConstraintStream<TestdataEntity> stream2 = constraintFactory.forEach(TestdataEntity.class);
+        assertThat(stream1).isSameAs(stream2);
```

Comment:
Without this PR, this would be failing.
However, with my changes to joiners, I managed to make it work.